sdk/trace/provider_test.go

```diff
@@ -80,6 +80,57 @@ func TestForceFlushAndShutdownTraceProviderWithoutProcessor(t *testing.T) {
 	assert.True(t, stp.isShutdown.Load())
 }
 
+func TestUnregisterFirst(t *testing.T) {
+	stp := NewTracerProvider()
+	sp1 := &basicSpanProcessor{}
+	sp2 := &basicSpanProcessor{}
+	sp3 := &basicSpanProcessor{}
+	stp.RegisterSpanProcessor(sp1)
+	stp.RegisterSpanProcessor(sp2)
+	stp.RegisterSpanProcessor(sp3)
+
+	stp.UnregisterSpanProcessor(sp1)
+
+	sps := stp.getSpanProcessors()
+	require.Len(t, sps, 2)
+	assert.Equal(t, sp2, sps[0].sp)
+	assert.Equal(t, sp3, sps[1].sp)
+}
+
+func TestUnregisterMiddle(t *testing.T) {
+	stp := NewTracerProvider()
+	sp1 := &basicSpanProcessor{}
+	sp2 := &basicSpanProcessor{}
+	sp3 := &basicSpanProcessor{}
+	stp.RegisterSpanProcessor(sp1)
+	stp.RegisterSpanProcessor(sp2)
+	stp.RegisterSpanProcessor(sp3)
+
+	stp.UnregisterSpanProcessor(sp2)
+
+	sps := stp.getSpanProcessors()
+	require.Len(t, sps, 2)
+	assert.Equal(t, sp1, sps[0].sp)
+	assert.Equal(t, sp3, sps[1].sp)
+}
+
+func TestUnregisterLast(t *testing.T) {
+	stp := NewTracerProvider()
+	sp1 := &basicSpanProcessor{}
+	sp2 := &basicSpanProcessor{}
+	sp3 := &basicSpanProcessor{}
+	stp.RegisterSpanProcessor(sp1)
+	stp.RegisterSpanProcessor(sp2)
+	stp.RegisterSpanProcessor(sp3)
+
+	stp.UnregisterSpanProcessor(sp3)
+
+	sps := stp.getSpanProcessors()
+	require.Len(t, sps, 2)
+	assert.Equal(t, sp1, sps[0].sp)
+	assert.Equal(t, sp2, sps[1].sp)
+}
```

Comment:
You have to use [`assert.Same`](https://pkg.go.dev/github.com/stretchr/testify/assert?utm_source=godoc#Same) instead of `assert.Equal`.

Currently, you can change e,g, `assert.Equal(t, sp2, sps[1].sp)` to  `assert.Equal(t, sp1, sps[1].sp)` and the test will still pass.